File: server/src/r2.ts
import {
  S3Client,
  PutObjectCommand,
  GetObjectCommand,
  HeadObjectCommand,
  DeleteObjectCommand,
  ListObjectsV2Command,
} from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { config } from './config';

export interface R2Config {
  client: S3Client;
  bucket: string;
}

export function itemKeyToObjectKey(itemKey: string, version: string): string {
  return `${itemKey.replace(/:/g, '/')}/${version}/_extract.tar.lz4`;
}

// Parses foo/bar/1700000000/_extract.tar.lz4 → { key: 'foo:bar', version: '1700000000' }
export function objectKeyToItemKey(objectKey: string): { key: string; version: string } | null {
  const m = objectKey.match(/^(.+)\/([^/]+)\/_extract\.tar\.lz4$/);
  if (!m) return null;
  return { key: m[1].replace(/\//g, ':'), version: m[2] };
}

export function r2ConfigForWorld(world: string): R2Config {
  const bucket = config.worlds[world];
  if (!bucket) throw new Error(`Unknown world: ${world}`);
  const client = new S3Client({
    region: 'auto',
    endpoint: `https://${config.r2.account_id}.r2.cloudflarestorage.com`,
    credentials: {
      accessKeyId: config.r2.access_key_id,
      secretAccessKey: config.r2.secret_access_key,
    },
  });
  return { client, bucket };
}

export async function presignedPutUrl(cfg: R2Config, objectKey: string, expiresInSeconds = 900): Promise<string> {
  return getSignedUrl(cfg.client, new PutObjectCommand({ Bucket: cfg.bucket, Key: objectKey }), { expiresIn: expiresInSeconds });
}

export async function presignedGetUrl(cfg: R2Config, objectKey: string, expiresInSeconds = 3600): Promise<string> {
  return getSignedUrl(cfg.client, new GetObjectCommand({ Bucket: cfg.bucket, Key: objectKey }), { expiresIn: expiresInSeconds });
}

export async function r2HeadObject(cfg: R2Config, objectKey: string): Promise<boolean> {
  try {
    await cfg.client.send(new HeadObjectCommand({ Bucket: cfg.bucket, Key: objectKey }));
    return true;
  } catch {
    return false;
  }
}

export async function r2DeleteObject(cfg: R2Config, objectKey: string): Promise<void> {
  await cfg.client.send(new DeleteObjectCommand({ Bucket: cfg.bucket, Key: objectKey }));
}

export async function r2ListObjects(cfg: R2Config, continuationToken?: string): Promise<{ keys: string[]; nextToken?: string }> {
  const resp = await cfg.client.send(new ListObjectsV2Command({
    Bucket: cfg.bucket,
    MaxKeys: 200,
    ContinuationToken: continuationToken,
  }));
  const keys = (resp.Contents ?? []).map(obj => obj.Key!).filter(Boolean);
  return { keys, nextToken: resp.NextContinuationToken };
}

File: c2server/internal/r2.go
package internal

import (
	"context"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

var objectKeyRegex = regexp.MustCompile(`^(.+)/([^/]+)/_extract\.tar\.lz4$`)

func ItemKeyToObjectKey(key, version string) string {
	return strings.ReplaceAll(key, ":", "/") + "/" + version + "/_extract.tar.lz4"
}

func ObjectKeyToItemKey(objectKey string) (key, version string, ok bool) {
	m := objectKeyRegex.FindStringSubmatch(objectKey)
	if m == nil {
		return "", "", false
	}
	return strings.ReplaceAll(m[1], "/", ":"), m[2], true
}

func newS3Client(world string) (*s3.Client, string, error) {
	bucket, ok := Cfg.Worlds[world]
	if !ok {
		return nil, "", fmt.Errorf("unknown world: %s", world)
	}
	client := s3.New(s3.Options{
		Region:       "auto",
		BaseEndpoint: aws.String(fmt.Sprintf("https://%s.r2.cloudflarestorage.com", Cfg.R2.AccountID)),
		Credentials:  credentials.NewStaticCredentialsProvider(Cfg.R2.AccessKeyID, Cfg.R2.SecretAccessKey, ""),
	})
	return client, bucket, nil
}

func NewBackupS3Client() *s3.Client {
	return s3.New(s3.Options{
		Region:       "auto",
		BaseEndpoint: aws.String(fmt.Sprintf("https://%s.r2.cloudflarestorage.com", Cfg.R2.AccountID)),
		Credentials:  credentials.NewStaticCredentialsProvider(Cfg.R2.AccessKeyID, Cfg.R2.SecretAccessKey, ""),
	})
}

func PresignedPutURL(world, objectKey string, expiry time.Duration) (string, error) {
	client, bucket, err := newS3Client(world)
	if err != nil {
		return "", err
	}
	presigner := s3.NewPresignClient(client)
	result, err := presigner.PresignPutObject(context.Background(), &s3.PutObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(objectKey),
	}, func(o *s3.PresignOptions) { o.Expires = expiry })
	if err != nil {
		return "", err
	}
	return result.URL, nil
}

func PresignedGetURL(world, objectKey string, expiry time.Duration) (string, error) {
	client, bucket, err := newS3Client(world)
	if err != nil {
		return "", err
	}
	presigner := s3.NewPresignClient(client)
	result, err := presigner.PresignGetObject(context.Background(), &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(objectKey),
	}, func(o *s3.PresignOptions) { o.Expires = expiry })
	if err != nil {
		return "", err
	}
	return result.URL, nil
}

func R2HeadObject(world, objectKey string) (bool, error) {
	client, bucket, err := newS3Client(world)
	if err != nil {
		return false, err
	}
	_, err = client.HeadObject(context.Background(), &s3.HeadObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(objectKey),
	})
	if err != nil {
		return false, nil
	}
	return true, nil
}

func R2DeleteObject(world, objectKey string) error {
	client, bucket, err := newS3Client(world)
	if err != nil {
		return err
	}
	_, err = client.DeleteObject(context.Background(), &s3.DeleteObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(objectKey),
	})
	return err
}

type ListResult struct {
	Keys      []string
	NextToken *string
}

func R2ListObjects(world string, continuationToken *string) (ListResult, error) {
	client, bucket, err := newS3Client(world)
	if err != nil {
		return ListResult{}, err
	}
	input := &s3.ListObjectsV2Input{
		Bucket:  aws.String(bucket),
		MaxKeys: aws.Int32(200),
	}
	if continuationToken != nil {
		input.ContinuationToken = continuationToken
	}
	resp, err := client.ListObjectsV2(context.Background(), input)
	if err != nil {
		return ListResult{}, err
	}
	keys := make([]string, 0, len(resp.Contents))
	for _, obj := range resp.Contents {
		if obj.Key != nil {
			keys = append(keys, *obj.Key)
		}
	}
	return ListResult{Keys: keys, NextToken: resp.NextContinuationToken}, nil
}
